sdks/go/pkg/beam/core/runtime/exec/sdf_test.go

```diff
@@ -0,0 +1,408 @@
+		tests := []struct {
+			name string
+			fn   *graph.DoFn
+			in   *FullValue
+			want *FullValue
+		}{
+			{
+				name: "SingleElem",
+				fn:   dfn,
+				in: &FullValue{
+					Elm: &FullValue{
+						Elm: &FullValue{
+							Elm:       3,
+							Elm2:      nil,
+							Timestamp: testTimestamp,
+							Windows:   testWindows,
+						},
+						Elm2: Restriction{5},
+					},
+					Elm2:      8.0,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+				want: &FullValue{
+					Elm:       8,
+					Elm2:      4,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+			},
+			{
+				name: "KvElem",
+				fn:   kvdfn,
+				in: &FullValue{
+					Elm: &FullValue{
+						Elm: &FullValue{
+							Elm:       3,
+							Elm2:      10,
+							Timestamp: testTimestamp,
+							Windows:   testWindows,
+						},
+						Elm2: Restriction{5},
+					},
+					Elm2:      18.0,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+				want: &FullValue{
+					Elm:       8,
+					Elm2:      12,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+			},
+		}
+		for _, test := range tests {
+			test := test
+			t.Run(test.name, func(t *testing.T) {
+				ctx := context.Background()
+				fake := &FakeNode{}
+				n := &ParDo{UID: 0, Fn: test.fn, Out: []Node{fake}}
+				node := ProcessSizedElementsAndRestrictions{PDo: n}
+
+				if err := node.Up(ctx); err != nil {
+					t.Fatalf("Up failed: %v", err)
+				}
+				if err := node.StartBundle(ctx, "bundle_id", DataContext{}); err != nil {
+					t.Fatalf("StartBundle failed: %v", err)
+				}
+
+				if err := node.ProcessElement(ctx, test.in); err != nil {
+					t.Fatalf("ProcessElement failed: %v", err)
+				}
+				got := fake.Vals[0]
+				if !cmp.Equal(got, test.want) {
+					t.Errorf("ProcessElement(%v) has incorrect output: got: %v, want: %v",
+						test.in, got, test.want)
+				}
+				if err := node.FinishBundle(ctx); err != nil {
+					t.Fatalf("FinishBundle failed: %v", err)
+				}
+				if err := node.Down(ctx); err != nil {
+					t.Fatalf("Down failed: %v", err)
+				}
+			})
+		}
+	})
+}
+
+// TestSplitAndSizeRestrictions verifies that the SplitAndSizeRestrictions node
+// both fulfills its contract and successfully invokes the accompanying SDF's
+// SplitRestriction and RestrictionSize methods to do so.
+func TestSplitAndSizeRestrictions(t *testing.T) {
+}
+
+// TestProcessSizedElementsAndRestrictions verifies that the
+// ProcessSizedElementsAndRestrictions node both fulfills its contract and
+// successfully invokes the accompanying SDF's CreateTracker and ProcessElement
+// methods to do so.
+func TestProcessSizedElementsAndRestrictions(t *testing.T) {
+}
+
+// FakeNode is used to capture the outputs of the node being tested by
+// TestSdfNodes. FakeNode appends each element that it receives in its
+// ProcessElement method to its Vals struct, which can then be read by the test.
+type FakeNode struct {
```

Comment:
Why write a new one rather than re-use CaptureNode? https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/unit_test.go#L28

Reply:
Didn't know that existed (or that whole file, really). Switched to it.